File: file_storage/internal/middleware/logging.go
package middleware

import (
	"file-storage/internal/di"
	"net/http"
	"strings"
	"time"

	"github.com/gorilla/mux"
	"go.uber.org/zap"
)

type responseWriter struct {
	http.ResponseWriter
	status int
	bytes  int
}

func (rw *responseWriter) Write(b []byte) (int, error) {
	n, err := rw.ResponseWriter.Write(b)
	rw.bytes += n
	return n, err
}
func (rw *responseWriter) WriteHeader(code int) {
	rw.status = code
	rw.ResponseWriter.WriteHeader(code)
}

func getIP(r *http.Request) string {
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		return strings.Split(xff, ",")[0]
	}
	return r.RemoteAddr
}

func LoggingMiddleware(logger *zap.Logger, config *di.Config) mux.MiddlewareFunc {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			start := time.Now()

			rw := &responseWriter{
				ResponseWriter: w,
				status:         http.StatusOK,
			}

			defer func() {
				if rec := recover(); rec != nil {
					duration := time.Since(start)
					template, _ := mux.CurrentRoute(r).GetPathTemplate()
					ctx := r.Context()
					logger.Error("panic recovered",
						zap.Any("panic", rec),
						zap.String("method", r.Method),
						zap.String("path", r.URL.Path),
						zap.String("route", template),
						zap.String("remote_addr", r.RemoteAddr),
						zap.Duration("duration", duration),
						zap.String("request_id", ctx.Value(config.RequestIDKey).(string)),
					)
					http.Error(rw, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
				}
			}()

			next.ServeHTTP(rw, r)

			duration := time.Since(start)
			template, _ := mux.CurrentRoute(r).GetPathTemplate()

			ctx := r.Context()
			fields := []zap.Field{
				zap.String("method", r.Method),
				zap.String("path", r.URL.Path),
				zap.String("route", template),
				zap.Int("status", rw.status),
				zap.Int("bytes", rw.bytes),
				zap.Duration("duration", duration),
				zap.String("remote_addr", getIP(r)),
				zap.String("request_id", ctx.Value(config.RequestIDKey).(string)),
			}

			switch {
			case rw.status >= 500:
				logger.Error("http request", fields...)
			case rw.status >= 400:
				logger.Warn("http request", fields...)
			default:
				logger.Info("http request", fields...)
			}
		})
	}
}

File: bandit_system/src/middleware/logging.py
import time
from contextvars import ContextVar
from typing import Optional

import structlog
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

from .request_id import get_request_id

# Context variable to store user UUID if available from auth
user_uuid_context: ContextVar[Optional[str]] = ContextVar("user_uuid", default=None)


def get_user_uuid() -> Optional[str]:
    """Get the current user UUID from context."""
    return user_uuid_context.get()


def set_user_uuid(user_uuid: str) -> None:
    """Set the user UUID in context."""
    user_uuid_context.set(user_uuid)


def get_ip(request: Request) -> str:
    """Extract the client IP address from the request."""
    xff = request.headers.get("X-Forwarded-For")
    if xff:
        return xff.split(",")[0].strip()
    return request.client.host if request.client else "unknown"


class LoggingMiddleware(BaseHTTPMiddleware):
    """
    Middleware that logs HTTP requests with details including:
    - Method, path, and route pattern
    - Status code and response size
    - Request duration
    - Client IP address
    - Request ID and user UUID (if available)

    Also handles panic recovery with logging.
    """

    def __init__(self, app, logger: structlog.BoundLogger):
        super().__init__(app)
        self.logger = logger

    async def dispatch(self, request: Request, call_next) -> Response:
        start_time = time.time()
        status_code = 500  # Default to 500 in case of uncaught exception
        response_size = 0

        try:
            # Process the request
            response = await call_next(request)
            status_code = response.status_code

            # Calculate response size if available
            if hasattr(response, "body"):
                response_size = len(response.body) if response.body else 0

            return response

        except Exception as exc:
            # Log panic/exception
            duration = time.time() - start_time
            route_path = request.url.path

            self.logger.error(
                "panic recovered",
                panic=str(exc),
                exception_type=type(exc).__name__,
                method=request.method,
                path=route_path,
                route=route_path,  # In FastAPI, we'd need route matching to get template
                remote_addr=get_ip(request),
                duration_ms=int(duration * 1000),
                request_id=get_request_id(),
                user_uuid=get_user_uuid(),
            )
            raise

        finally:
            # Log the request
            duration = time.time() - start_time
            route_path = request.url.path

            # Log at appropriate level based on status code
            if status_code >= 500:
                self.logger.error(
                    "http request",
                    method=request.method,
                    path=route_path,
                    route=route_path,
                    status=status_code,
                    bytes=response_size,
                    duration_ms=int(duration * 1000),
                    remote_addr=get_ip(request),
                    request_id=get_request_id(),
                    user_uuid=get_user_uuid(),
                )
            elif status_code >= 400:
                self.logger.warning(
                    "http request",
                    method=request.method,
                    path=route_path,
                    route=route_path,
                    status=status_code,
                    bytes=response_size,
                    duration_ms=int(duration * 1000),
                    remote_addr=get_ip(request),
                    request_id=get_request_id(),
                    user_uuid=get_user_uuid(),
                )
            else:
                self.logger.info(
                    "http request",
                    method=request.method,
                    path=route_path,
                    route=route_path,
                    status=status_code,
                    bytes=response_size,
                    duration_ms=int(duration * 1000),
                    remote_addr=get_ip(request),
                    request_id=get_request_id(),
                    user_uuid=get_user_uuid(),
                )